### ISSUE TITLE
Prompt user to run "dolt login" on PermissionDenied error on "dolt push" 

### ISSUE THREAD

**Issue by oscarbatori** [dolthub/dolt#783]
This at least eliminates the case where they have not yet authenticated. We should also indicate that if they have already authenticated some more digging might be needed.

**Comment by 7vikpeculiar** [comment]
Can I work on this ?
If yes, can you please tell how to suggest

**Comment by reltuk** [comment]
This might be a great first start at diving into the dolt code base, and we would totally welcome an improvement here.

There's a question about how to get the right behavior in a non-intrusive way and in as many places as it might be helpful.

Here's some notes about how it fits together currently:

* The dolt commands that interact with remotes construct a `DoltDB` referring to the remote URL and do various things with it, like `Clone`, `Pull`, etc. When those operations fail, the underlying RPC error is embedded in the error that gets returned. So each command could be extended to check for a `PermissionDenied` error and add the suggestion text if appropriate.
    * I think the push failure error is here: https://github.com/liquidata-inc/dolt/blob/master/go/cmd/dolt/commands/push.go#L302
    * I think a clone failure error would be here: https://github.com/liquidata-inc/dolt/blob/master/go/cmd/dolt/commands/clone.go#L291

For the underlying error, https://godoc.org/google.golang.org/grpc/status#Code can be used to extract the error code. `codes.PermissionDenied` is the one we're interested in: https://godoc.org/google.golang.org/grpc/codes#Code

Currently at those calls sites, whether the underlying operation is using credentials isn't known. Setup for accessing the remote database goes through a factory here:
* https://github.com/liquidata-inc/dolt/blob/master/go/libraries/doltcore/dbfactory/grpc.go#L66
which gets its connection parameters from an environment object that includes user-level and repo-level configuration:
* https://github.com/liquidata-inc/dolt/blob/master/go/libraries/doltcore/env/environment.go#L622
So that environment is returning the connection options that include the per-call credentials if they are appropriate for accessing the targeted remote.

The call site that's dealing with the returned error doesn't currently know whether the environment ended up using credentials on the outbound request that got the `PermissionDenied` or not. This is the most impactful part of any attempt to fix this currently. We could explicitly model that knowledge somewhere. It could be on the `DoltDB` instance itself, perhaps, and be ultimately populated by those `dbfactory` methods that are responsible for creating the remotes. Or perhaps it could be queried by the command implementations at the point of the error – the environment and the remote we were attempting to communicate with include enough information to know if we would use credentials on the request (might be race-y under state changes, but that might not matter a whole lot for this use case). We could also use a client interceptor to wrap the error on its way back to the command implementation so that it could know. That might look something like...

```go
type doltRemoteError struct {
	s *status.Status
	isUncredentialed bool
}

func (dre *doltRemoteError) GRPCStatus() *status.Status {
	return dre.s
}

func (dre *doltRemoteError) Error() string {
	return dre.s.Err().Error()
}

func IsUncredentialedPermissionsError(err error) bool {
        if status.Code(err) != codes.PermissionDenied {
                return false
        }
	if dre, ok := err.(*doltRemoteError); ok {
		return dre.isUncredentialed
	}
	return false
}

func doltRemoteErrorClientIntercetor(credentialed bool) grpc.UnaryClientInterceptor {
	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
		err := invoker(ctx, method, req, reply, cc, opts...)
		if err != nil {
			if s, ok := status.FromError(err); ok {
				return &doltRemoteError{s, !credentialed}
			}
		}
		return nil
	}
}
```

I'm not exactly sure about the best way to wrap errors, and `status.Status` errors, in the current ecosystem, but something like that would get a method at the call site that could be used. I think an alternative is to add some client details protobuf message type and use `status.Status.WithDetails()` to thread the information through. Or a `ptypes.Struct` with some well defined string field names...which seems not great.

In either case, then `DoltEnv.GetGRPCDialParams` would put that client interceptor on the dial options with https://godoc.org/google.golang.org/grpc#WithChainUnaryInterceptor.

Alternatively, that interceptor could avoid the new error type and look something like:

```go
func doltRemoteErrorClientIntercetor(credentialed bool) grpc.UnaryClientInterceptor {
	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
		err := invoker(ctx, method, req, reply, cc, opts...)
		if err != nil && !credentials && status.Code(err) == codes.PermissionDenied {
			return status.Error(codes.PermissionDenied, "permission denied: try running `dolt login` to generate dolt credentials for use with dolthub")
		}
		return err
	}
}
```

where the suggestion just goes into the `err` message that the status error is carrying. That's not beautiful and doesn't allow for much control over the output or internationalization, etc.

If the err itself carries enough information know know whether we hit this case or not, then the right place to add the messaging might not be in the commands themselves, like push, pull and clone, but in the errhand/verr stuff. In particular, `AddCause(err error)` could choose add appropriate usage information when it runs into this. That's kind of here:
* https://github.com/liquidata-inc/dolt/blob/master/go/cmd/dolt/errhand/derr.go#L138

Anyway...a bit of a brain dump. Please feel free to follow up with any questions you might have here.

Thanks for your interest in dolt :).

**Comment by 7vikpeculiar** [comment]
I am a bit confused, I thought that when there is an error during push, we suggest the user to check if `dolt login` is run or not...

**Comment by reltuk** [comment]
Yes, certainly that is the intention. The above description of how we might go about it was based on not wanting to make the suggestion if the user is already using credentials – in that case, `dolt login` is probably not the right thing to do.

**Comment by 7vikpeculiar** [comment]
I am beginner on this project, would you mind breaking the steps into chunks, for me

Thanks a lot

**Comment by zachmu** [comment]
I think as an easier first step, it's enough to examine the error returned from the push operation, see if it is a 403, and ask "have you run `dolt login`?"

We'll get some false positives there, but it doesn't need to be perfect.

**Comment by mucleck** [comment]
Hi! I've tried to solve the error. Im junior PHP developer learning Go so sorry if this is not the best solution. 

So the first thing I did was to check why handlePushError could not detect that the incoming error was a ErrUnknowPushErr. This is because the detection doesnt know that the incomming error is a wrapped one. The  solution is to check with error.Is(). 

<img width="1147" height="495" alt="Image" src="https://github.com/user-attachments/assets/6937c2b2-2c77-4088-8aca-b307c5d4b468" />

Changed to:

<img width="1073" height="515" alt="Image" src="https://github.com/user-attachments/assets/82f7e256-9913-47e0-a610-c9b115ee212a" />

Now the second problema is that the error that is envolved is not an error its a string. I changed this here

<img width="1344" height="364" alt="Image" src="https://github.com/user-attachments/assets/968cd180-0f47-4b3e-91a4-3ff2914c0597" />

Here I get lost so any explanation of why this code that Gemini generated works would be apreciated. The only thing I understand is that it needs error.As() to check if any of the wrapped errors is from the specified type. 

<img width="1107" height="510" alt="Image" src="https://github.com/user-attachments/assets/366a6087-b865-4af9-983d-6c57d2946f3c" />

<img width="770" height="351" alt="Image" src="https://github.com/user-attachments/assets/1681c677-b5aa-4109-9ee5-85a05bd3fa42" />

Any hint for this part?
Am I close to the solution?
 Ty!!

**Comment by Adarshrai24** [comment]
I’d like to work on this with a simple approach: detect PermissionDenied errors and add a hint suggesting dolt login. This may show the hint even when credentials are present, but it should improve the common unauthenticated case

**Comment by kevglynn** [comment]
Hi — picking this up. The hint text already exists in `handlePushError` but is unreachable due to two bugs in the error-matching path (sentinel compared with `==` on a wrapped error, and the gRPC status flattened to a string before the `PermissionDenied` check). Planning to fix those so the existing hint fires, per @zachmu's guidance of keeping it simple. Will open a PR shortly.
